packages/lib/BaseApplication.ts

```diff
@@ -481,14 +481,14 @@ export default class BaseApplication {
 		// appLogger.debug('Reducer action', this.reducerActionToString(action));
 
 		const result = next(action);
-		const newState = store.getState();
 		let refreshNotes = false;
 		let refreshFolders: boolean | string = false;
 		// let refreshTags = false;
 		let refreshNotesUseSelectedNoteId = false;
 		let refreshNotesHash = '';
 
 		await reduxSharedMiddleware(store, next, action);
+		const newState = store.getState();
```

Comment:
Why was it changed?

Reply:
It's the fix of a latent bug. Since reduxSharedMiddleware() is async and sometimes invokes plugin code in it, the value of store.getState() aquired before its invocation is sometimes out-of-date after its execution is finished.

In the current implementation, the modification of the state in reduxSharedMiddleware() is lost. To use the latest state, I moved the line below.

In this pull request case, the bug is revealed. The following sequence is its example.
```
    01 BaseApplication.ts/generalMiddleware() 
        (Setting.value('notes.sortOrder.field') === 'title')
    02   reduxSharedMiddleware.ts/reduxSharedMiddleware()
    03     eventManager.ts/appStateEmit()
    04       perNotebookSortOrder.ts/onNotebookSelectionMayChange()
             (Setting.setValue('notes.sortOrder.field', 'user_updated_time'))
    05     eventManager.ts/appStateEmit()
    06   reduxSharedMiddleware.ts/reduxSharedMiddleware()
    07 BaseApplication.ts/generalMiddleware() 
        (The setting at 01 is used and the modification of the setting at 04 disappears.)
```

Beside generalMiddleware(), I also felt bad smells in reduxSharedMiddleware.ts/reduxSharedMiddleware(). But I did't review the code.

Reply:
Additionally, the same type of smells is in app-mobile/root.tsx/generalMiddleware().

Reply:
Yes that indeed makes sense. But on the other hand, since this code was like this for a long time and is not unit tested, could you do a quick check on the code below that uses newState to make sure it still works as expected?

Reply:
> could you do a quick check on the code below that uses newState to make sure it still works as expected?

When I implemented this, I checked the code in generalMiddleware() and judged the move is not problematic. On the other hand,  I didn't review reduxSharedMiddleware.ts/reduxSharedMiddleware() or app-mobile/root.tsx/generalMiddleware().

For reduxSharedMiddleware(), at a glance, it seems for me that moving `const newState = store.getState();` afterward is appropriate. For app-mobile/root.tsx/generalMiddleware(), I cannot judge about app-mobile due to lack of knowledge.

> this code was like this for a long time

There is a big warn here. Since appStateEmit() invokes plugin code through joplin.workspace.onNoteSelectionChange() API, any third party code can be executed here. Just because it has been safe in the past does not mean it will be safe in the future.